NMS-15158: Add flow table (#5622)
OpenNMS/opennms: docs/modules/operation/pages/deep-dive/flows/introduction.adoc

[[ga-flow-support-introduction]]
= Flows

Flows are summaries of network traffic sent by network devices (switches, routers, and so on).
This information includes, but is not limited to, source and destination addresses, source and destination ports, octet count, and duration of activity.

Collecting and analyzing flows data provides a picture of network usage and helps to diagnose issues.
Persisting flows for long-term storage can aid in forensic analysis.

{page-component-title} provides the following:

* A platform to collect, persist, and visualize flows, with support for NetFlow versions 5 and 9, IPFIX, and sFlow.
* Inventory enrichment (mapping to {page-component-title} nodes).
* Application classification.
* Horizontal scaling.
* Enterprise reporting.

> * Top K statistics by interface, application, host, conversation with QoS.

Reviewer: ```suggestion
* Top K statistics by interface, application, host, and conversation with QoS.
```

* Collect flows data in a xref:deep-dive/flows/distributed.adoc#flows-remote[distributed/remote network] (add a Minion).
* Process a xref:deep-dive/flows/sentinel/sentinel.adoc#flows-scaling[large volume of flows data] (add Sentinel to scale).
* Issues with flows at scale and queries taking too long (add the OpenNMS https://github.com/OpenNMS/nephron[streaming analytics tool]).

== How it works

At a high level, with a xref:deep-dive/flows/basic.adoc#flows-basic[basic setup], {page-component-title} processes flows as follows:

* <<deep-dive/telemetryd/introduction.adoc#ga-telemetryd, Telemetryd>> receives and decodes flows.
* Telemetryd adapters convert the flows to a canonical flow model.
* Flows are enriched:
** The <<deep-dive/flows/classification-engine.adoc#ga-flow-support-classification-engine, classification engine>> tags flows and groups them under a name based on a set of rules.
** Metadata related to associated nodes (such as IDs and categories) are added to the flows.
* Enriched flows are persisted in Elasticsearch and/or forwarded to Kafka.
* (Optional) The OpenNMS streaming analytics tool aggregates flows and outputs them to Elasticsearch, Cortex, or Kafka.

> * (Optional) The REST API generates summaries and time series data from the flows or aggregates stored on Elasticsearch.
> * OpenNMS Helm visualizes flows and aggregates:

Author: ```suggestion
. (Optional) The REST API generates summaries and time series data from the flows or aggregates stored on Elasticsearch.
. OpenNMS Helm visualizes flows and aggregates:
```

> ** The Flow Deep Dive dashboard visualizes flows and aggregates that are stored in Elasticsearch using a flows datasource.

Reviewer: Is it ES using the flows datasource or the dashboard? If the latter, suggest:

```suggestion
** The Flow Deep Dive dashboard uses a flows datasource to visualize flows and aggregates that are stored in Elasticsearch.
```
I think it's probably the former, but we should reword the sentence to make that clearer. 
Maybe:

The Flow Deep Dive dashboard visualizes flows and aggregates that are stored in Elasticsearch from a flows datasource.

Reviewer 2: I suggest:
The Flow Deep Dive dashboard visualizes flows and aggregates, retrieving their data from Elasticsearch via a flows datasource.

Author: ```suggestion
** The Flow Deep Dive dashboard visualizes flows and aggregates, retrieving their data from Elasticsearch via a flows datasource.
```

> ** The Cortex Flow Deep Dive dashboard visualizes aggregates that are stored in Cortex using a Prometheus datasource.

Reviewer: Same question as line 42. Which is using the datasource?

```suggestion
** The Cortex Flow Deep Dive dashboard uses a Prometheus datasource to visualize aggregates that are stored in Cortex.
```

If the former:

The Cortex Flow Deep Dive dashboard visualizes aggregates that are stored in Cortex from a Prometheus datasource.
